core/src/avm2/verify.rs

```diff
@@ -0,0 +1,1126 @@
+                            )
+                        })?;
+                }
+
+                _ => {}
+            }
+
+            i += 1;
+        }
+
+        if !save_current_work {
+            worklist.pop();
+        }
+    }
+
+    Ok(())
+}
+
+fn optimize<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    method: &BytecodeMethod<'gc>,
+    code: &mut Vec<Op>,
+    jump_targets: HashSet<i32>,
+) {
+    // These make the code less readable
+    #![allow(clippy::manual_filter)]
+    #![allow(clippy::single_match)]
+
+    // This can probably be done better by recording the receiver in `Activation`,
+    // but this works since it's guaranteed to be set in `Activation::from_method`.
+    let this_value = activation.local_register(0);
+
+    let this_class = if let Some(this_class) = activation.subclass_object() {
+        if this_value.is_of_type(activation, this_class.inner_class_definition()) {
+            Some(this_class)
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    // Initial set of local types
+    let mut local_types = vec![None; method.body().unwrap().num_locals as usize];
+    local_types[0] = this_class;
+
+    // Invalidate local types if they should be invalidated
+    for op in &*code {
+        match op {
+            Op::SetLocal { index }
+            | Op::Kill { index }
+            | Op::IncLocal { index }
+            | Op::IncLocalI { index }
+            | Op::DecLocal { index }
+            | Op::DecLocalI { index } => {
+                if (*index as usize) < local_types.len() {
+                    local_types[*index as usize] = None;
+                }
+            }
+            Op::HasNext2 {
+                object_register,
+                index_register,
+            } => {
+                if (*object_register as usize) < local_types.len() {
+                    local_types[*object_register as usize] = None;
+                }
+                if (*index_register as usize) < local_types.len() {
+                    local_types[*index_register as usize] = None;
+                }
+            }
+            _ => {}
+        }
+    }
+
+    let mut previous_op = None;
+    for (i, op) in code.iter_mut().enumerate() {
+        if let Some(previous_op_some) = previous_op {
+            if !jump_targets.contains(&(i as i32)) {
+                match op {
+                    Op::CoerceB => match previous_op_some {
+                        Op::CoerceB
+                        | Op::Equals
+                        | Op::GreaterEquals
+                        | Op::GreaterThan
+                        | Op::LessEquals
+                        | Op::LessThan
+                        | Op::PushTrue
+                        | Op::PushFalse
+                        | Op::Not
+                        | Op::IsType { .. }
+                        | Op::IsTypeLate => {
+                            previous_op = Some(op.clone());
```

Comment:
We could write this as `previous_op = mem::replace(op, Op::Nop)`, but I am not sure if that is really worth it.

Reply:
I think this current implementation is more readable.